### PR TITLE
feat(push): device-token GET/DELETE + registry-backed health/debug (PR 2)

### PR DESCRIPTION
**Pull request by murdore**
## PR 2 of multi-device push — registration API (GET/DELETE) + registry-backed health/debug

Stacked on #96 (PR 1). Adds the multi-device registration surface on top of the SQLite store. **Dual-writes** to the registry AND the legacy `device-tokens.json` so `/api/notify` keeps working until the PR-5 fan-out cutover — no notify behavior change yet.

### Changes
- **`POST /api/device-token`** — also upserts into `device_tokens` (accepts `deviceId`/`deviceName`/`appEnv` from newer apps); keeps the legacy JSON dual-write + `process.env.DEVICE_TOKEN`; `TOKENS_DIR` now uses `shooterDataDir()`.
- **`GET /api/device-token`** — auth-gated; lists active devices with **masked tokens** + per-platform counts.
- **`DELETE /api/device-token {id}`** — auth-gated soft-delete by registry row id.
- **`/api/health`** — `hasDeviceToken` now reflects the registry (env var kept as fallback); authenticated response gains `devices:{ios,android,total}`; warning reworded to "No devices registered".
- **`/api/debug`** — device token sourced from the most-recent active iOS row; adds `registeredDevices` counts.
- **`device-format.ts`** (`maskToken` + `toDeviceListItem`) + 4 unit tests.

### Verified end-to-end against the running server (curl, not unit tests)
Ran the real `tsx server.ts` in an isolated `SHOOTER_HOME`/port and drove every endpoint:
- Registered **two iOS + one Android** device → `GET` returns all 3 (`counts:{ios:2,android:1}`), masked tokens, no raw `token` field — the case that previously overwrote.
- Token rotation (same `deviceId`, new token) → same row, no duplicate.
- `DELETE` by id → device drops from the active list (`is_active=0` on disk, auditable).
- **Restart** → registry persists (SQLite, not in-memory).
- Probes: no-auth `401`, missing `id` `400`, missing token `400`, invalid platform `400`, delete-nonexistent `removed:0`.

### Scoped deviations (kept tight; no `gen:types` churn)
Device counts live in a new `devices` field rather than mutating the generated `HealthChecks`/`HealthConfiguration` types. `config-guard.ts`/`config.yaml` `deviceId` deferred to PR 8 (config UI, where it's consumed).

Gate: `pnpm test` (36 push tests + all suites) · `pnpm check` 0/0 · `pnpm build` · ESLint/Prettier — all green.

## Summary by CodeRabbit

* **New Features**
  * Added multi-device push notification support; users can now register and manage multiple devices per account.
  * Device registration accepts optional metadata (device name, bundle ID, environment selection).
  * New device management API with masked token display for security.
  * Health and debug endpoints now report registered device counts by platform.

* **Tests**
  * Added test coverage for device registration and formatting functionality.

### CODE REVIEW

**Comment by Tara-ag** [comment]
## 🔒 Security Review Summary

I've completed a security-focused review of this PR. Here are my findings:

### ✅ Critical Security Checks - PASSED

| Check | Status | Notes |
|-------|--------|-------|
| **Authentication** | ✅ Pass | All endpoints (`GET`, `DELETE`, `POST`) use `validateAuth()` with `timingSafeEqual` |
| **SQL Injection** | ✅ Pass | Parameterized queries via better-sqlite3; dynamic placeholders are programmatically generated |
| **Token Masking** | ✅ Pass | Raw tokens never exposed in GET responses; console logs only token length |
| **Path Traversal** | ✅ Pass | Uses `shooterDataDir()` with proper path joining |
| **Transaction Safety** | ✅ Pass | `upsert()` uses `this.db.transaction()` for atomicity |

### 🔍 Key Security Highlights

1. **DELETE /api/device-token**: Properly validates auth, validates `id` is a non-empty string, returns 404 for unknown ids (safe information disclosure posture)

2. **Token Storage**: SQLite with WAL mode; tokens stored securely with soft-delete for audit trail

3. **Migration Safety**: File import errors are caught and logged without crashing; files renamed after processing to prevent re-import loops

4. **Identity Theft Guard**: The `upsert()` logic properly guards against token stealing via the `WHERE` clause conditions on conflict

### 💡 Non-blocking Suggestions

- Consider adding rate limiting to device registration endpoints in a future PR (defense in depth)
- The `isDeviceRecord` type guard was strengthened in response to earlier feedback - good iteration

**Overall**: This PR introduces no new security vulnerabilities and follows the project's established secure patterns for a publicly-exposed tunnel service.

File: src/lib/modules/server/push/device-token-store.ts [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+        bundle_id     TEXT,
+        registered_at TEXT NOT NULL,
+        last_seen_at  TEXT NOT NULL,
+        failure_count INTEGER NOT NULL DEFAULT 0,
+        is_active     INTEGER NOT NULL DEFAULT 1,
+        UNIQUE(token),
+        UNIQUE(device_id, platform)
+      );
+      CREATE INDEX IF NOT EXISTS idx_device_tokens_active
+        ON device_tokens(platform, is_active);
```
Comment (Tara-ag) [inline src/lib/modules/server/push/device-token-store.ts:75]: 💡 **MINOR: Consider adding an index for `app_env` filtering**

The `listActiveForEnv()` method queries by `platform = ? AND app_env = ? AND is_active = 1`, but the current index `idx_device_tokens_active` only covers `(platform, is_active)`. For large device registries, this may result in a table scan when filtering by app_env.

Consider adding a composite index:
```sql
CREATE INDEX IF NOT EXISTS idx_device_tokens_env 
  ON device_tokens(platform, app_env, is_active);
```

This aligns with the indexing pattern used elsewhere in the codebase (e.g., `terminal-store.ts`).

File: src/lib/modules/server/push/device-token-store.ts [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+    // Normalize at the store boundary so a caller passing an un-trimmed token
+    // can't create a logically-duplicate row (whitespace) that would poison
+    // fan-out/pruning. The HTTP route already trims, but the store shouldn't
+    // trust every caller.
+    const token = input.token.trim();
+    if (token.length === 0) {
+      throw new Error('DeviceTokenStore.upsert: token must be a non-empty string');
+    }
+
+    this.db.transaction(() => {
```
Comment (Tara-ag) [inline src/lib/modules/server/push/device-token-store.ts:209]: 💡 **MINOR: Consider validating `deviceId` length/format**

The `deviceId` from user input is trimmed but has no length limit or format validation before being used in SQL queries. While SQLite handles this safely, consider adding a reasonable length limit (e.g., 255 chars) to prevent potential abuse with extremely long identifiers.

```typescript
if (deviceId !== null && deviceId.length > 255) {
  throw new Error('DeviceTokenStore.upsert: deviceId exceeds maximum length');
}
```

File: src/routes/api/device-token/+server.ts [hunk 3]
```diff
@@ -63,8 +132,24 @@ export const POST: RequestHandler = async ({ request }) => {
     return json({ error: 'Missing device token (deviceToken or token)' }, { status: 400 });
```
Comment (Tara-ag) [inline src/routes/api/device-token/+server.ts:152]: 💡 **MINOR: Consider rate limiting for device registration**

The POST endpoint accepts device registrations without any rate limiting. Given this is a public-facing endpoint (via Cloudflare Tunnel), consider adding basic rate limiting to prevent abuse - e.g., max N registrations per IP per minute.

This is consistent with the security focus for other endpoints that handle push notification targets.

File: src/lib/types/device.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+  sent: number;
+  success: boolean;
+  timestamp: string;
+}
+
+/** Consecutive delivery failures before a token is considered dead. */
+export const MAX_FAILURE_COUNT = 3;
+
+/** Runtime guard for a DeviceRecord coming from an untrusted boundary. */
+export function isDeviceRecord(v: unknown): v is DeviceRecord {
```
Comment (Tara-ag) [inline src/lib/types/device.ts:109]: ✅ **Good: Proper type guard with comprehensive validation**

The `isDeviceRecord` type guard correctly validates all required fields including the nullable string fields (`bundleId`, `deviceId`, `friendlyName`). This is important for runtime safety when deserializing from the database.

Note: The CodeRabbit comment about strengthening the guard appears to have been addressed - the current implementation properly validates all DeviceRecord fields.

File: src/routes/api/device-token/+server.ts [hunk 5]
```diff
@@ -63,8 +132,24 @@ export const POST: RequestHandler = async ({ request }) => {
...
   }
```
Comment (Tara-ag) [inline src/routes/api/device-token/+server.ts:153]: 💡 **MINOR: `deviceName` field could benefit from length validation**

The `deviceName` (mapped to `friendlyName`) is trimmed but has no maximum length validation. Consider adding a reasonable limit (e.g., 100 characters) to prevent UI rendering issues and database bloat:

```typescript
const friendlyName = typeof body.deviceName === 'string' 
  ? body.deviceName.trim().slice(0, 100) || null 
  : null;
```

File: src/lib/modules/server/push/device-token-store.ts [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+      throw new Error('DeviceTokenStore.upsert: row not found after write');
+    }
+    return rec;
+  }
+
+  private importFile(filePath: string, renameSuffix: string, now: Date): void {
+    if (!fs.existsSync(filePath)) {
+      return;
+    }
+    try {
```
Comment (Tara-ag) [inline src/lib/modules/server/push/device-token-store.ts:282]: 💡 **MINOR: `importFile` error handling could log more context**

When `importFile` catches an error during migration, it logs the error but doesn't include the file path in the warning message. Consider enhancing the log:

```typescript
console.warn(`[device-token-store] failed to import ${filePath}:`, err instanceof Error ? err.message : String(err));
```

This helps operators diagnose which specific file caused issues during startup migration.

File: src/routes/api/debug/+server.ts [hunk 7]
```diff
@@ -30,6 +38,10 @@ export const GET: RequestHandler = ({ request }) => {
     },
     environment: env.NODE_ENV || 'development',
     hasApiKey: !!env.API_KEY,
+    registeredDevices: {
```
Comment (Tara-ag) [inline src/routes/api/debug/+server.ts:41]: 💡 **MINOR: Consider caching `listActiveForEnv` result for `registeredDevices`**

The debug endpoint calls `deviceTokenStore.listActiveForEnv('ios', apnsEnv)` and then separately calls `deviceTokenStore.listActive('android')`. The iOS result is stored in `iosDevices` but then `listActive('ios')` is called again for `registeredDevices.ios`.

Consider reusing the already-fetched `iosDevices`:
```typescript
registeredDevices: {
  android: deviceTokenStore.listActive('android').length,
  ios: iosDevices.length,  // reuse instead of re-querying
},
```

File: src/routes/api/health/+server.ts [hunk 8]
```diff
@@ -36,11 +37,19 @@ export const GET: RequestHandler = ({ request, url }) => {
...
 
+  // Multi-device registry counts (replace the single-token env check).
+  const iosDeviceCount = deviceTokenStore.listActive('ios').length;
+  const androidDeviceCount = deviceTokenStore.listActive('android').length;
+  const registeredDeviceCount = iosDeviceCount + androidDeviceCount;
+
   const checks: HealthChecks = {
     hasApiKey: !!env.API_KEY?.trim(),
     hasAPNsConfig: !!(env.APNS_KEY_ID?.trim() && env.APNS_TEAM_ID?.trim() && env.APNS_KEY?.trim()),
     hasBundleId: !!env.APNS_BUNDLE_ID?.trim(),
```
Comment (Tara-ag) [inline src/routes/api/health/+server.ts:48]: ✅ **Good: Backward-compatible health check**

The `hasDeviceToken` check maintains backward compatibility by checking both the new registry AND the legacy env vars. This ensures smooth migration without breaking existing monitoring that depends on this field.

The new `devices` field provides richer information for authenticated callers while preserving the existing contract.

File: tests/device-token-store.test.cjs [hunk 9]
```diff
@@ -0,0 +1,423 @@
+/**
```
Comment (Tara-ag) [inline tests/device-token-store.test.cjs:1]: ✅ **Good: Comprehensive test coverage**

Excellent test coverage with 32 tests covering:
- Basic CRUD operations
- Token rotation scenarios
- Platform-specific filtering
- Migration from legacy formats
- Edge cases (empty tokens, non-existent IDs)
- Identity theft protection
- Soft-delete behavior

The use of isolated temp directories and injected timestamps follows the established patterns in the codebase.

File: src/lib/modules/server/push/device-format.ts [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+ * `maskToken` keeps a short prefix + suffix and hides the middle so the
+ * registered-devices list (and any log of it) never carries a full APNs/FCM
+ * push token. `toDeviceListItem` maps an internal DeviceRecord to the masked
+ * API shape returned by GET /api/device-token — deliberately dropping the raw
+ * `token` field.
+ */
+
+import type { DeviceListItem, DeviceRecord } from '$lib/types';
+
+export function maskToken(token: null | string | undefined): string {
```
Comment (Tara-ag) [inline src/lib/modules/server/push/device-format.ts:13]: ✅ **Good: Secure token masking implementation**

The `maskToken` function correctly handles edge cases:
- Returns empty string for null/undefined/empty input
- Fully masks tokens shorter than 8 characters (prevents near-full exposure)
- Shows appropriate prefix/suffix for longer tokens
- Real APNs (64 hex) and FCM (~152) tokens get meaningful masking

This prevents accidental logging of full push tokens in the device list API.

File: src/routes/api/device-token/+server.ts [hunk 11]
```diff
@@ -25,20 +30,84 @@ function readTokens(): { android?: string; ios?: string } {
...
+  // Old apps omit appEnv → match the server's configured APNs gateway.
+  return env.APNS_PRODUCTION === 'true' ? 'production' : 'sandbox';
+}
+
 function writeTokens(tokens: { android?: string; ios?: string }): void {
   if (!existsSync(TOKENS_DIR)) {
     mkdirSync(TOKENS_DIR, { mode: 0o700, recursive: true });
   }
   writeFileSync(TOKENS_FILE, JSON.stringify(tokens, null, 2), { encoding: 'utf-8', mode: 0o600 });
 }
```
Comment (Tara-ag) [inline src/routes/api/device-token/+server.ts:52]: ✅ **Good: Proper auth on all HTTP methods**

All three HTTP methods (GET, POST, DELETE) correctly call `validateAuth(request)` before processing. This maintains the security boundary for the device registry endpoints.

The DELETE endpoint also properly validates the request body and returns appropriate HTTP status codes (400 for bad input, 404 for not found).

File: package.json [hunk 12]
```diff
@@ -76,7 +76,7 @@
     "format:generated": "prettier --write ./src/lib/types/generated/*",
```
Comment (Tara-ag) [inline package.json:76]: ✅ **Good: Test script properly extended**

The test script correctly adds the two new test files (`device-token-store.test.cjs` and `device-format.test.cjs`) to the chain. The `&&` chaining ensures all tests run and the exit code aggregates failures.

**Review by Tara-ag — APPROVED** [review]
## Review Summary

**Files reviewed:** 11
**New issues raised:** 6 minor suggestions
**Blocking issues:** None

### Assessment

This PR implements the multi-device push notification registry (PR 2 of the stack) with solid engineering practices:

**Security ✅**
- All endpoints (`GET`, `POST`, `DELETE` /api/device-token) properly authenticate via `validateAuth()`
- Push tokens are masked in API responses via `maskToken()` - no secret leakage
- Uses parameterized SQL queries throughout - no injection vulnerabilities
- Proper file permissions (0o600 for tokens file, 0o700 for directory)

**Architecture ✅**
- Clean separation: `DeviceTokenStore` for persistence, `device-format.ts` for presentation
- SQLite with WAL mode follows established patterns (terminal-store, etc.)
- Global singleton keyed on `globalThis` for module loader consistency
- Dual-write strategy maintains backward compatibility during migration

**Testing ✅**
- 32 comprehensive unit tests for the store covering edge cases
- 4 tests for token masking functionality
- Isolated temp directories prevent test pollution
- Injected timestamps enable deterministic testing

**Type Safety ✅**
- Hand-written types in `src/lib/types/device.ts` (following decision.ts pattern)
- Proper barrel export via `$lib/types`
- Runtime type guard `isDeviceRecord` validates all fields including nullable strings

### Minor Suggestions (Non-blocking)

1. **Index optimization**: Consider adding `(platform, app_env, is_active)` composite index for `listActiveForEnv()` efficiency
2. **Input validation**: Consider max length limits for `deviceId` and `deviceName`/`friendlyName`
3. **Rate limiting**: Consider basic rate limiting on the POST endpoint for abuse prevention
4. **Debug endpoint**: Could reuse `iosDevices` result instead of re-querying for `registeredDevices.ios`
5. **Error logging**: Migration error messages could include the file path for easier debugging

### Existing Comments

The CodeRabbit comments about `DeviceRecord` schema alignment and `isDeviceRecord` validation appear to have been addressed in the current revision.

**Recommendation:** Approve - this is a well-crafted PR that follows project conventions and maintains security boundaries.

**Comment by murdore** [comment]
Addressed the **legacy token sink synchronization** finding on the DELETE handler (`cb52e87`). The DELETE path is now symmetric with POST: after a successful delete it re-derives `device-tokens.json` + `process.env.DEVICE_TOKEN` from the most-recently-seen surviving registry entry (or clears them when none remain), so a just-deleted token can't keep receiving pushes through the legacy `/api/notify` path during the transition window. Runtime-verified end-to-end (register 2 → delete 1 → sink re-points to survivor → delete last → sink cleared). The #100 cutover later removes the legacy sinks entirely.

File: src/routes/api/device-token/+server.ts [hunk 13]
```diff
@@ -25,20 +36,98 @@ function readTokens(): { android?: string; ios?: string } {
...
+  }
+
+  let body: unknown;
+  try {
+    body = await request.json();
+  } catch {
+    return json({ error: 'Invalid JSON body' }, { status: 400 });
+  }
+
+  const id =
```
Comment (Tara-ag) [inline src/routes/api/device-token/+server.ts:85]: ⚠️ **MINOR: Consider validating `id` format in DELETE endpoint**

The `id` parameter is trimmed but doesn't validate the UUID format before passing to `deleteById()`. While this is not a security issue (SQLite handles it safely), validating the format would provide clearer error messages for malformed input:

```typescript
const UUID_REGEX = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
if (!UUID_REGEX.test(trimmedId)) {
  return json({ error: 'Invalid id format' }, { status: 400 });
}
```

This aligns with the existing input validation patterns in the POST endpoint.

File: src/routes/api/debug/+server.ts (hunk 7 above)
Comment (Tara-ag) [inline src/routes/api/debug/+server.ts:41]: ⚠️ **MINOR: Inconsistent filtering in `registeredDevices` counts**

The `registeredDevices.ios` count uses `iosDevices.length` which is filtered by `apnsEnv`, but `registeredDevices.android` uses `listActive('android')` without environment filtering. This creates an inconsistency where iOS counts are environment-scoped but Android counts are not.

Consider either:
1. Adding environment filtering for Android (if FCM also has environment separation)
2. Using `listActive('ios').length` for iOS to match the Android approach

Current code:
```typescript
registeredDevices: {
  android: deviceTokenStore.listActive('android').length,
  ios: iosDevices.length,  // filtered by apnsEnv
},
```

**Comment by Tara-ag** [comment]
## Security-Focused Review Complete 🔒

I've completed a security-centric review of this multi-device push notification registry PR. Here are my findings:

### Blocking Criteria Verification

| Criterion | Result |
|-----------|--------|
| Hardcoded secrets | ✅ None found |
| Command/shell injection | ✅ None found (parameterized SQL) |
| Auth bypass | ✅ All endpoints protected |
| Secret leakage | ✅ Tokens masked, files 0o600 |
| Path traversal | ✅ Uses `shooterDataDir()` helper |

### Security Highlights

1. **Authentication**: All three HTTP methods (`GET`, `POST`, `DELETE`) on `/api/device-token` correctly call `validateAuth()` before processing.

2. **SQL Injection Prevention**: The `DeviceTokenStore` uses parameterized queries exclusively. Dynamic placeholder construction in `pruneByTokens()` and `touchLastSeen()` is safe (uses `?` placeholders, not string concatenation).

3. **Secret Masking**: The `maskToken()` function properly prevents full push token exposure in API responses—critical for a public Cloudflare Tunnel deployment.

4. **Input Boundaries**: Defensive length limits (512/256 chars) prevent row bloat attacks without affecting legitimate clients.

### Recommendation

**Approve** — No new blocking issues identified. The existing minor suggestions from @Tara-ag (index optimization, UUID validation, rate limiting considerations) are appropriate follow-ups for future PRs but do not block this merge.

The PR demonstrates solid security hygiene for a feature that handles sensitive push tokens and exposes new API surface area.

**Comment by Tara-ag** [comment]
## Autonomous Review Summary

**Reviewer:** Open-Thinking Model  
**Files Analyzed:** 11  
**New Issues Raised:** 0  

### Security Assessment ✅

| Area | Status | Notes |
|------|--------|-------|
| Authentication | ✅ | All endpoints (`GET`, `POST`, `DELETE`) correctly validate Bearer tokens via `validateAuth()` |
| SQL Injection | ✅ | All queries use parameterized statements; dynamic placeholders built safely |
| Secret Handling | ✅ | Tokens masked in responses (`maskToken`), 0o600 file permissions, no raw token logging |
| Path Safety | ✅ | Uses `shooterDataDir()` with `path.join()` - no traversal risk |
| Input Validation | ✅ | Field length limits enforced (MAX_TOKEN_LENGTH, etc.) with 400 responses |

### Key Observations

**Strengths:**
- Comprehensive test coverage (36 tests across 2 new test files)
- Follows established patterns (SQLite+WAL, globalThis singleton, barrel exports)
- Clean dual-write strategy for backward compatibility during migration
- Proper transaction handling with rollback on failure
- Token masking prevents secret leakage in API responses

**Already Covered by Previous Reviewers:**
- DeviceRecord schema alignment (CodeRabbit - resolved)
- `isDeviceRecord` validation completeness (CodeRabbit - resolved)
- Index optimization suggestion for `listActiveForEnv()` (Tara-ag)
- UUID format validation for DELETE endpoint (Tara-ag)
- Debug endpoint device count inconsistency (Tara-ag)

### Verdict

**APPROVE** - No blocking issues identified. The PR demonstrates solid security practices and follows project conventions. All substantive concerns have been appropriately raised by previous reviewers.

---
*This is an autonomous review conducted following the project standards in CLAUDE.md and the blocking criteria for security-critical code.*

File: src/routes/api/device-token/+server.ts [hunk 14]
```diff
@@ -25,20 +36,98 @@ function readTokens(): { android?: string; ios?: string } {
...
   const authError = validateAuth(request);
   if (authError) {
     return authError;
   }
 
-  let body: { bundleId?: string; deviceToken?: string; platform: string; token?: string };
+  let body: {
+    appEnv?: string;
+    bundleId?: string;
+    deviceId?: string;
```
Comment (Tara-ag) [inline src/routes/api/device-token/+server.ts:125]: 💡 **MINOR: Direct mutation of `process.env.DEVICE_TOKEN` may not persist across requests**

The DELETE handler sets `process.env.DEVICE_TOKEN = iosFallback` directly. However, `process.env` mutations in Node.js are string-only and don't reliably persist across the process lifetime, especially if the env was originally loaded from a `.env` file via dotenv.

Since this is a dual-write transition mechanism (noted in the comment about PR-5 fan-out cutover), consider whether this mutation is necessary, or if the legacy path should simply read from the registry instead of relying on the env var staying in sync.

If the mutation is intentional for backward compatibility, consider adding a comment explaining why this is safe in this specific context.

File: tests/device-format.test.cjs [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+const HEX64 = 'a1b2c3d4e5f60718293a4b5c6d7e8f90a1b2c3d4e5f60718293a4b5c6d7e8f90';
+
+runTest('Test 1: maskToken shows a short prefix + suffix and hides the middle', () => {
+  const masked = maskToken(HEX64);
+  assertEqual(masked, 'a1b2c3…8f90', 'masked form');
+  assertTrue(!masked.includes(HEX64.slice(10, 50)), 'middle of the token is not present');
+  assertTrue(masked.length < HEX64.length, 'masked string is shorter than the raw token');
+});
+
+runTest('Test 2: maskToken masks short tokens too (no near-full reveal)', () => {
```
Comment (Tara-ag) [inline tests/device-format.test.cjs:63]: 💡 **MINOR: Missing test coverage for `< 8` character token masking**

The `maskToken` function has a safety branch that returns `'••••'` for tokens shorter than 8 characters (lines 16-20 in device-format.ts), but this branch is not exercised by the current tests. Consider adding a test case to ensure this safety behavior is locked in:

```javascript
runTest('Test 5: maskToken fully masks very short tokens (< 8 chars)', () => {
  assertEqual(maskToken('abcd'), '••••', '4-char token fully masked');
  assertEqual(maskToken('abcdefg'), '••••', '7-char token fully masked');
});
```

This aligns with the bootstrapped standards pattern of guarding all branches in security-sensitive masking functions.

File: src/lib/modules/server/push/device-token-store.ts [hunk 16]
```diff
@@ -0,0 +1,354 @@
...
+   */
+  migrate(now: Date = new Date()): void {
+    this.importFile(path.join(this.dataDir, 'device-tokens.json'), '.migrated', now);
+    this.importFile(path.join(this.dataDir, 'device-token-seeds.json'), '.processed', now);
+  }
+
+  /** Soft-delete a set of dead tokens (lazy pruning after a failed delivery). */
+  pruneByTokens(tokens: readonly string[]): number {
+    if (tokens.length === 0) {
+      return 0;
```
Comment (Tara-ag) [inline src/lib/modules/server/push/device-token-store.ts:130]: ✅ **Good: Safe parameterized query construction in `pruneByTokens` and `touchLastSeen`**

The dynamic `IN` clause construction using `tokens.map(() => '?').join(', ')` is safe because:
1. The `?` placeholders are not user-controlled - they're generated from the array length
2. The actual token values are passed as separate parameters to `.run(...tokens)`
3. better-sqlite3 handles proper parameter binding, preventing SQL injection

This follows the same safe pattern used elsewhere in the codebase (e.g., terminal-store.ts).

**Review by Tara-ag — APPROVED** [review]
## Review Summary

**Files reviewed:** 11
**New issues raised:** 2 minor suggestions
**Blocking issues:** None

### Assessment

This PR implements the multi-device push notification registry (PR 2 of the stack) with solid engineering practices:

**Security ✅**
- All endpoints (`GET`, `POST`, `DELETE` /api/device-token) properly authenticate via `validateAuth()`
- Push tokens are masked in API responses via `maskToken()` - no secret leakage
- Uses parameterized SQL queries throughout - no injection vulnerabilities
- Proper file permissions (0o600 for tokens file, 0o700 for directory)
- Safe dynamic IN clause construction with proper parameter binding

**Architecture ✅**
- Clean separation: `DeviceTokenStore` for persistence, `device-format.ts` for presentation
- SQLite with WAL mode follows established patterns (terminal-store, etc.)
- Global singleton keyed on `globalThis` for module loader consistency
- Dual-write strategy maintains backward compatibility during migration

**Testing ✅**
- 33 comprehensive unit tests for the store covering edge cases
- 4 tests for token masking functionality
- Isolated temp directories prevent test pollution
- Injected timestamps enable deterministic testing

**Type Safety ✅**
- Hand-written types in `src/lib/types/device.ts` (following decision.ts pattern)
- Proper barrel export via `$lib/types`
- Runtime type guard `isDeviceRecord` validates all fields including nullable strings

### New Issues Raised (Non-blocking)

1. **process.env mutation concern**: The DELETE handler directly mutates `process.env.DEVICE_TOKEN`. While this is part of the dual-write transition strategy, consider whether this mutation is necessary or if the legacy path should read from the registry instead.

2. **Missing test coverage**: The `maskToken` function's safety branch for tokens < 8 characters is not exercised by current tests.

### Existing Comments Acknowledged

The existing comments from CodeRabbit and Tara-ag cover:
- Index optimization for `listActiveForEnv()`
- Input validation length limits
- Rate limiting considerations
- Debug endpoint query reuse
- JSDoc documentation

All previously raised blocking issues appear to have been addressed.

**Recommendation:** Approve - this is a well-crafted PR that follows project conventions and maintains security boundaries.

**Comment by murdore** [comment]
Superseded by #104, which consolidated this entire stack into a single squashed commit and was **merged into `release`** (merge commit `a9919f2`).

All changes from this PR are included in #104 — https://github.com/juspay/shooter/pull/104

Closing as superseded. The branch is retained for now (not deleted).
